rethinkdb/datadog_checks/rethinkdb/metrics/_base.py

```diff
@@ -0,0 +1,146 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import logging
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    Generic,
+    Iterator,
+    List,
+    Literal,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    TypedDict,
+    TypeVar,
+    Union,
+)
+
+from datadog_checks.base import AgentCheck
+
+from ..connections import Connection
+from ..queries import QueryEngine
+from ..types import Metric, MetricType
+from ..utils import dotted_join, lookup_dotted, to_timestamp
+
+logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+DocumentT = TypeVar("DocumentT", bound=Mapping)
+
+ModifierName = Literal['total', 'ok_warning', 'timestamp']
+TotalModifier = TypedDict('TotalModifier', {'name': Literal['total'], 'map': Callable[[Any], Sequence[T]]})
+Modifier = Union[ModifierName, TotalModifier]
+
+MetricDefinition = TypedDict(
+    'MetricDefinition',
+    {
+        'type': MetricType,
+        'path': str,  # Used as the default name.
+        'name': str,  # An explicit name for the metric.
+        'modifier': Optional[Modifier],
+    },
+    total=False,
+)
+
+Enumeration = TypedDict('Enumeration', {'path': str, 'index_tag': str, 'metrics': List[MetricDefinition]})
+
+
+class DocumentMetricCollector(Generic[DocumentT]):
+    """
+    TODO(before-merging): Explain how to use this.
```

Comment:
?

Reply:
🤣 Yup, still have to address this. Just turned stale as I moved the refactor to a different PR for clarity: #6092